src/modules/did/did.service.ts

```diff
@@ -495,4 +517,71 @@ export class DIDService implements OnModuleInit, OnModuleDestroy {
       this.logger.debug(inspect(jobsCounts, { depth: 2, colors: true }));
     }
   }
+
+  private async getChangedIdentities(
+    fromBlock: number,
+    topBlock: number
+  ): Promise<string[]> {
+    const didEventFilters = [
+      this.didRegistry.filters.DIDAttributeChanged(null),
+      this.didRegistry.filters.DIDDelegateChanged(null),
+      this.didRegistry.filters.DIDOwnerChanged(null),
+    ];
+    const events = [];
+    while (fromBlock < topBlock) {
+      const toBlock = Math.min(
+        topBlock,
+        fromBlock + this.MAX_EVENTS_QUERY_INTERVAL
+      );
+
+      const intervalEvents = (
+        await Promise.all(
+          didEventFilters.map((filter) =>
+            this.didRegistry.queryFilter(filter, fromBlock, toBlock)
+          )
+        )
+      ).flat();
+      this.logger.debug(
+        `Fetched ${intervalEvents.length} DID events from interval [${fromBlock}, ${toBlock}]`
+      );
+      events.push(...intervalEvents);
+      fromBlock += this.MAX_EVENTS_QUERY_INTERVAL;
+    }
+    this.logger.debug(`Update document events count ${events.length}`);
+    const changedIdentities = events.map((event) => {
+      return event.args.identity;
+    });
+    // Deduplicate identities if they are appears in multiple events
+    return [...new Set(changedIdentities).values()];
+  }
+
+  private async markStaleDocuments() {
+    const syncs = await this.latestDidSyncRepository.find({
+      order: { createdDate: 'DESC' },
+    });
+    const fromBlock = syncs.length > 0 ? syncs[0].block : 0;
+    const topBlock = await this.provider.getBlockNumber();
+    this.logger.debug(
+      `Getting DID update events from block ${fromBlock} to block ${topBlock}...`
+    );
+    const changedIdentities = await this.getChangedIdentities(
+      fromBlock,
+      topBlock
+    );
+    const staleDIDs = (
+      await this.didRepository.find({ select: ['id'] })
+    ).filter((doc) => {
+      const identity = addressOf(doc.id);
+      return changedIdentities.includes(identity);
+    });
+    await this.didSyncStatusRepository
+      .createQueryBuilder()
+      .useTransaction(true)
+      .update(DidSyncStatusEntity)
+      .set({ status: DidSyncStatus.Stale })
+      .where({ document: In(staleDIDs) })
+      .execute();
```

Comment:
I seems to me like a better approach would be to rely on the database to do the filtering in the `where` clause, rather than reading all dids into application memory and then filtering in the application. I think that, if the number of DIDs became large enough, this `this.didRepository.find({ select: ['id']` query would probably fail.

I suppose the challenge might be how to use the addresses returned from `changedIdentities` to find the `didSyncStatus` entities to update. I think a couple options are:
1. Store the address of the identity on the `DidSyncStatusEntity`
2. Join to `DIDDocumentEntity` in the query and filter on the `id` property. Of course the addresses would need to be transformed to `did:ethr` DID ids.

Note that I think that option 2 might be slightly less complicated if we got rid of the one-to-one join entity and just added a new column to the DidDocumentEntity.
I think we only really need a single `isStale` boolean column (or could also keep the enum), so the impact on the performance of the database query when reading the DID Document would be  negligible, I think.

Reply:
Thanks for suggestions. Option 2 seems more safe to me, because it excludes case when two status entities connected to the same document. I would also like to avoid changing document entity. It is used everywhere in ssi-hub and I am bit afraid to cause some unexpected changes. And besides property `isState` is not inherently part of the document.